ntp-daemon/src/peer.rs

```diff
@@ -1,177 +1,281 @@
+use std::{pin::Pin, sync::Arc};
+
 use tracing::warn;
 
 use ntp_proto::{
-    IgnoreReason, NtpClock, NtpHeader, NtpInstant, Peer, PeerSnapshot, ReferenceId, SystemConfig,
-    SystemSnapshot,
+    AcceptSynchronizationError, IgnoreReason, NtpClock, NtpHeader, NtpInstant, NtpTimestamp, Peer,
+    PeerSnapshot, ReferenceId, SystemConfig, SystemSnapshot,
 };
 use ntp_udp::UdpSocket;
-use tracing::instrument;
+use tracing::{info, instrument};
+
+use tokio::{
+    net::ToSocketAddrs,
+    sync::watch,
+    time::{Instant, Sleep},
+};
+
+/// Only messages from the current reset epoch are valid. The system's reset epoch is incremented
+/// (with wrapping addition) on every reset. Only after a reset does the peer update its reset
+/// epoch, thereby indicating to the system that the reset was successful and the peer's messages
+/// are valid measurements again.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub struct ResetEpoch(u64);
+
+impl ResetEpoch {
+    #[must_use]
+    pub const fn inc(mut self) -> Self {
+        self.0 = self.0.wrapping_add(1);
+
+        self
+    }
+}
 
-use tokio::{net::ToSocketAddrs, sync::watch, time::Instant};
+#[derive(Debug, Clone, Copy)]
+pub struct PeerIndex {
+    pub index: usize,
+}
 
 #[derive(Debug, Clone, Copy)]
 pub enum MsgForSystem {
     /// Received a Kiss-o'-Death and must demobilize
-    MustDemobilize,
-    /// There is no measurement available, either because no
-    /// packet has been received yet, or because synchronization was rejected
-    NoMeasurement,
+    MustDemobilize(PeerIndex),
     /// Received an acceptable packet and made a new peer snapshot
-    Snapshot(PeerSnapshot),
-}
-
-pub struct PeerChannels {
-    pub peer_snapshot: watch::Receiver<MsgForSystem>,
-    pub peer_reset: watch::Receiver<u64>,
+    Snapshot(PeerIndex, ResetEpoch, PeerSnapshot),
 }
 
-#[instrument(skip(clock, config, system_snapshots, reset))]
-pub async fn start_peer<A: ToSocketAddrs + std::fmt::Debug, C: 'static + NtpClock + Send>(
-    addr: A,
+struct PeerTask<C> {
```

Comment:
I'm bothered by this struct. It's just a random collection of fields, not really an abstraction. It is hard to figure out when certain pieces of state are updated. I also don't really see a good way to fix that though, so I'd love your suggestions